Cascading policy initial implementation

pmm-sumo opened the pull request:
**Description:** 

This is initial work on cascading policy implementation, where several rules can be applied with separate sampling rates. A refactoring will most likely follow to unify the other policy behaviours

There's a top level max rate of spans per second and separate rates for each of the rules. Additionally, some rules can specify the max span rate as `-1`, which means that whatever bandwidth is available can be accommodated by them, after all traces are evaluated.

For example, if there's a set of rules that specify following:

```
            spans_per_second: 300,
            rules: [
              {
                spans_per_second: 150,
                numeric_attribute: {key: key1, min_value: 50, max_value: 100}
              },
              {
                spans_per_second: -1
              }
            ]
```

Then traces with total up to 150 spans matching `key1 between 50 and 100` will be sampled and the remainder will be available for all traces.

Lets say that we have following traces:

* `Trace A` with 100 spans, with one of them having `key1=70`
* `Trace B` with 100 spans, with one of them having `key1=80`
* `Trace C` with 100 spans, neither having matching `key1` attribute
* `Trace D` with 100 spans, neither having matching `key1` attribute
* `Trace E` with 100 spans, neither having matching `key1` attribute

In such case, the processor might yield following traces:
* `Trace A` (for the rule specified)
* `Trace C`
* `Trace D`

`Trace B` would be skipped because there's not enough bandwidth available for the rule. `Trace C` and `Trace D` would be selected because they could fit the remaining 200 spans.

**Link to tracking Issue:** N/A

**Testing:** Unit test for cascading policy added. Manual tests of solution done

**Documentation:** Example addded

sumo-drosiek commented on processor/samplingprocessor/tailsamplingprocessor/metrics.go:
```diff
@@ -25,9 +25,9 @@ import (
-	tagPolicyKey, _    = tag.NewKey("policy")
-	tagSampledKey, _   = tag.NewKey("sampled")
-	tagSourceFormat, _ = tag.NewKey("source_format")
+	tagPolicyKey, _       = tag.NewKey("policy")
+	tagSampledKey, _      = tag.NewKey("sampled")
+	tagSourceFormat, _    = tag.NewKey("source_format")
```
I suppose it should be left as it was

piotr-sumo commented:
Good question: as @sumo-drosiek mentioned, will E, D and C saturate all bandwidth?

piotr-sumo commented on processor/samplingprocessor/tailsamplingprocessor/processor.go:
```diff
@@ -281,9 +339,12 @@ func (tsp *tailSamplingSpanProcessor) ConsumeTraceData(ctx context.Context, td c
 			case sampling.Pending:
 				// All process for pending done above, keep the case so it doesn't go to default.
+			case sampling.SecondChance:
+				// It shouldn't normally get here, keep the case so it doesn't go to default, like above.
```
maybe we should log on WARN level when something like that happens?

piotr-sumo commented on processor/samplingprocessor/tailsamplingprocessor/sampling/cascading.go:
```diff
@@ -0,0 +1,192 @@
+type cascadingPolicy struct {
+	logger *zap.Logger
+
+	currentSecond        int64
```
`int64` should be one space after `currentSecond`

pmm-sumo replied:
GoFmt rules state that a section needs to be aligned to the longest string

piotr-sumo commented on processor/samplingprocessor/tailsamplingprocessor/sampling/cascading.go:
```diff
@@ -0,0 +1,192 @@
+	}
+}
+
+func (cre *cascadingRuleEvaluation) updateRate(currSecond int64, trace *TraceData) Decision {
```
I really like these small functions

pmm-sumo commented:
@sumo-drosiek 
> What if traces arrive in reverse order?
> 
> Do we yield `E`, `D`, `C` ?

> Good question: as @sumo-drosiek mentioned, will E, D and C saturate all bandwidth?

This are great questions :) The way it works is that there are two phases of sampling. The first phase will fit the rules with defined max rate up to the available rate (hence either `A` or `B` would be selected)

The second phase ("second chance" - really looking for a better name) is then reiterating over the rest of the traces and selects those. 

So, to answer the original question - no. If batches are like in the example, then always either `A` or `B` will be present in the yielded collection of traces

pmm-sumo commented on processor/samplingprocessor/tailsamplingprocessor/config/config.go:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-package tailsamplingprocessor
+package config
```
I moved it to another package to avoid dependency cycle

sumo-drosiek replied:
Do we want to create PR for the upstream or not really?

pmm-sumo replied:
I would love to bring this work upstream, but it's a complex change that need further refactoring, lets see

pmm-sumo commented:
Made some fixes, added `duration` policy and refactored the code a bit. Also, it now has more observability features (metrics) which make it easier to understand what's going on:

```
otelcol_count_cascading_traces_sampled{considered="matched",rule="duration",sampled="false"} 718
otelcol_count_cascading_traces_sampled{considered="matched",rule="duration",sampled="true"} 39
otelcol_count_cascading_traces_sampled{considered="not-matched",rule="duration",sampled="false"} 18440
otelcol_count_cascading_traces_sampled{considered="second_chance",rule="",sampled="false"} 18330
otelcol_count_cascading_traces_sampled{considered="second_chance",rule="",sampled="true"} 110
otelcol_count_cascading_traces_sampled{considered="second_chance",rule="everything_else",sampled="second_chance"} 18440
```
